12525 move web2print to bundle

mattamon opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.5`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.5` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #12525 

## Additional info  
Moving web to print functionality into PimcoreWebToPrint
Able to install and uninstall via Installer. Permissions are in the Bundle
Moved translations into the bundle
Moved documents printabstract, printcontainer and printpage to bundle

mattamon committed:
Moving web2print to bundle

mattamon committed:
Renaming Web2PrintBundle to PimcoreWeb2PrintBundle
Moving Migration into CoreBundle
Starting to add Document class loader

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php
#	bundles/AdminBundle/templates/admin/index/index.html.twig
#	bundles/CoreBundle/config/pimcore/default.yaml
#	bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php
#	bundles/Web2PrintBundle/public/js/web2print.js
#	composer.json
#	models/DataObject/ClassDefinition/Data/AdvancedManyToManyObjectRelation.php
#	models/DataObject/ClassDefinition/Data/AdvancedManyToManyRelation.php
#	models/DataObject/ClassDefinition/Data/Classificationstore.php

mattamon committed:
Updating bundle to refactoring tips

mattamon committed:
Adding upgrade notes
Reverting routing changes to document controllers

mattamon committed:
Remove document data loader

mattamon committed:
Moving messenger
Changed fixed document routes

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

mattamon committed:
Changing routing for save and remove from session

mattamon committed:
Change bundle namespace

mattamon committed:
Update tests

mattamon committed:
Fix test errors

mattamon committed:
Add type

mattamon commented on bundles/CoreBundle/src/Migrations/Version20220829132224.php:
```diff
@@ -75,8 +75,8 @@ private function migrateCustomViewsPerspectives(string $fileName, string $scope)
         $configs = $this->loadLegacyConfigs('web2print.php');
-        if (count($configs) > 0) {
-            $web2PrintConfigs = Config::getWeb2PrintConfig();
+        if (count($configs) > 0 && class_exists('Pimcore\Bundle\WebToPrintBundle\Config')) {
```
I am not to sure about changing this migration

kingjia90 replied:
~I'd wrap the entire `function up()` with a IF block that checks if it exists under the SettingStore and in case "skips" everything.
The `down()` is empty anyways.~
Nevermind, i've misunderstood the issue.

mattamon replied:
I am not sure how this check should look like. Do you have an example?

kingjia90 replied:
```
public function up(Schema $schema): void
{
  if  (!SettingsStore::get('web_to_print', 'pimcore_web_to_print'){
      $this->migrateCommonConfigurations('predefined-properties.php', 'pimcore_predefined_properties');
       ...
       $this->migrateWeb2PrintSettings();
  }
}
```
when that exists, it should do nothing, assuming the migration was already done in the past or doesn't need because there are no more legacy config files.

mattamon committed:
Change category to web2print

mattamon committed:
Update docs and deleting not used files

mattamon commented on lib/Perspective/Config.php:
```diff
@@ -132,7 +132,7 @@ public static function getStandardPerspective(): array
                     'document' => [
                         'items' => [
-                            'addPrintPage' => \Pimcore\Config::getWeb2PrintConfig()['enableInDefaultView'] ? true : false, // hide add print documents by default
+                            'addPrintPage' => \Pimcore\Bundle\WebToPrintBundle\Config::getWeb2PrintConfig()['enableInDefaultView'] ? true : false, // hide add print documents by default
```
Maybe we should refactor this in another way? Only add this when the bundle is enabled and installed?

mattamon replied:
I removed this completely since it will be show if the bundle is installed.

kingjia90 replied:
Please see https://github.com/pimcore/pimcore/pull/14175#pullrequestreview-1293804387, this config `enableInDefaultView` should be just removed.

kingjia90 commented on bundles/WebToPrintBundle/translations/admin.en.yaml:
```diff
@@ -0,0 +1,24 @@
+web2print_pdf_conversion: PDF Conversion
+web2print_saving_pdf_document: Save PDF Document
+web2print_author: Author
+web2print_printermarks: Printermarks
```
some web2print aliases went missing after https://github.com/pimcore/pimcore/pull/14069
eg.
https://github.com/pimcore/pimcore/blob/6b7b4ee668fa7c5907c26d068eee20ca28f4388d/bundles/CoreBundle/translations/aliases.json#L32-L33

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221220152444.php:
```diff
@@ -0,0 +1,46 @@
+        if(!SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle', 'pimcore')) {
+            SettingsStore::set('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle', true, 'bool', 'pimcore');
+        }
+
```
Need also to update the existing permissions `web2print_settings` to be in its own category, as in the installer

mattamon replied:
This is true, we should do this for all the other bundles too right?

kingjia90 replied:
yes, it should be like that, maybe let's ask everybody to double-check the PR they merged

kingjia90 commented on bundles/AdminBundle/public/js/pimcore/document/document.js:
```diff
@@ -115,7 +117,7 @@ pimcore.document.document = Class.create(pimcore.element.abstract, {
 
             Ext.Ajax.request({
-                url: Routing.getBaseUrl() + "/admin/" + this.getType() + '/save?task=' + task,
+                url: Routing.generate(this.getSaveRoute()),
```
`?task` param is missing here, leading to [problems](https://github.com/pimcore/pimcore/blob/68e86786778387b23aa20de74a7812d84a8688b6/bundles/AdminBundle/src/Controller/Admin/Document/DocumentControllerBase.php#L381)

kingjia90 reviewed (changes requested):
Found some minor issues after a quick check, but overall seems working fine

kingjia90 commented:
One input but unrelated to the changes in the PR is: how should it behave when you remove/deactivate the web2print bundle and you still have printpage and printcontainers? What about a cleanup tool or a "archived/invalid documents" tree node and/or filter to hide them? And the generated PDFs and job file? Just wondering if uninstaller should take care of this or is up to the devs

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	bundles/WebToPrintBundle/src/Controller/Document/PrintpageControllerBase.php

mattamon committed:
Adding task param
Adding missing translations
Adding category to permissions in migration

mattamon commented:
I was wondering about this too, how we should handle those files, because with this PR we do not even have a fallback for the printpage or printcontainer in the core. 

@brusch wdyt? what should happen with printcontainer or printpage documents after the bundle gets uninstalled or disabled?

mattamon committed:
Moving add functionality to bundle

mattamon committed:
Adding db installer script
Adding uninstall documentation

kingjia90 commented:
There are some occurences of the printpage and printcontainer that should be extracted

- [ ] `printcontainer` and `printpage`, maybe put in a pushable array property instead of inline array https://github.com/pimcore/pimcore/blob/ac85677346262376585b33a4fc210ad808d69b4f/bundles/AdminBundle/public/js/pimcore/document/document.js#L643-L647
- [x] https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/CoreBundle/config/pimcore/default.yaml#L206
- [x] https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/AdminBundle/src/Controller/Admin/Document/DocumentController.php#L285-L289 to be removed since it's on bundles/WebToPrintBundle/src/Controller/Document/PrintpageControllerBase.php::addAction()
- [x] to remove https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/AdminBundle/public/js/pimcore/document/tree.js#L403
- [ ] https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/AdminBundle/public/js/pimcore/document/tree.js#L445-L453 and https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/AdminBundle/public/js/pimcore/document/tree.js#L465-L473
- [ ] https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/AdminBundle/public/js/pimcore/settings/docTypes.js#L147
- [x] this should be added through install/uninstall https://github.com/pimcore/pimcore/blob/13f1dc1d660963c63807835bd8832198d5f9f5cf/bundles/InstallBundle/dump/install.sql#L99

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

mattamon committed:
Remove printpage and printcontainer occurencnes
Adapt installer

mattamon commented:
@kingjia90 for the other removals, those are hardcoded elements, so we need to find a way to make it configurable.

kingjia90 commented on bundles/WebToPrintBundle/src/Installer.php:
```diff
@@ -26,16 +28,34 @@ class Installer extends SettingsStoreAwareInstaller
     {
         $this->installDatabaseTable();
+        $enums = array_merge(self::STANDARD_DOCUMENT_ENUM_TYPES, self::BUNDLE_EXTRA_DOCUMENT_ENUM_TYPES);
+        $this->modifyEnumTypes($enums);
```
if there's an another bundle that also adds new document types, it won't work properly, or at least it would all depend on the installation order/priority, right?

For instance, headlessdocuments
https://github.com/pimcore/headless-documents/blob/bc6884c65ef3c30ea516afc5327226e96feeffed/src/Installer.php#L56

Could we instead use something that adds/remove items from existing array?

mattamon replied:
You are totally right, I will check that.

mattamon replied:
But then we have to change it for the headless documents too right? 
Because if you install it, printpage and printcontainer will be added never the less?

kingjia90 replied:
yes, i've opened a issue for that https://github.com/pimcore/headless-documents/issues/111

mattamon committed:
Getting existing enums for installer

mattamon committed:
Remove code duplicate

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

mattamon committed:
Add WebToPrint Installer to Unittests

mattamon committed:
Adding return types

mattamon committed:
Kick print page unit test

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	bundles/CoreBundle/translations/admin.en.yaml

mattamon committed:
Change !empty to count() > 0

mattamon committed:
Remove check

mattamon committed:
Adding doctrine compatibility layer

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php
#	doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

kingjia90 commented on bundles/WebToPrintBundle/src/Installer.php:
```diff
@@ -0,0 +1,121 @@
+            $type = ArrayParameterType::STRING;
+        }
+        $db = Db::get();
+        $db->executeQuery('ALTER TABLE documents MODIFY COLUMN `type` ENUM(:enums);', ['enums' => $enums], ['enums' => $type]);
```
If one uninstalls, the document *type* would be wiped for the printpage and printcointainer documents (because the enum doesn't exist anymore).
The documentation is clear about it, you have to clean up first before uninstalling.
https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/doc/Development_Documentation/18_Tools_and_Features/30_Rendering_PDFs.md?plain=1#L14-L18


Howerer, it could be that one just want to temporary uninstall the bundle to re-install later or if someone miss these instructions/notes (which even happenend to myself 😅 ) in the documentation, then it will end up with documents with empty types and this `DELETE FROM documents WHERE type IN ('printcontainer', 'printpage'); ` won't even work retroactively.

I would suggest to move this ENUM removal out of the uninstall, instead display a warning message on uninstall, something like "if you want to clean up the web2print data, please re-install the bundle and run bin/console web2print:cleanup before uninstalling again` and in this command we do: 

DELETE FROM documents WHERE type IN ('printcontainer', 'printpage'); 
ALTER TABLE documents MODIFY COLUMN `type` ENUM...
DROP TABLE documents_printpage; 


wdyt?

mattamon replied:
Ah yes, this makes totally sense. Unfortunately we should really move this to it's own command. Somehow adding new documents, is still not that easy as it should be though. 
I will try to add the warning in the uninstall and update the documentation and add the command. Maybe we should not make the command web2print specific, more like "clean up documents"?
So you can use it for all types of documents you want to clean up?

kingjia90 commented on lib/Twig/Extension/DocumentHelperExtensions.php:
```diff
@@ -53,18 +57,27 @@ public function getTests(): array
             }),
         ];
+
+        if(class_exists(PimcoreWebToPrintBundle::class)) {
```
We should extract this part to the bundle as well. 
It would likely just need to be registered as a service
https://github.com/pimcore/pimcore/pull/7531/files

If valid, please also specify that these would be only available when web2print bundle is installed. https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/doc/Development_Documentation/02_MVC/02_Template/02_Template_Extensions/README.md?plain=1#L87-L89

kingjia90 commented on tests/Model/Document/DocumentTest.php:
```diff
@@ -367,16 +367,4 @@ public function testDocumentSerialization(): void
         $this->assertEquals(count($document->getEditables()), count($loadedDocument->getEditables()));
     }
-
-    public function testDocumentPrint(): void
```
what's the reason behind this `testDocumentPrint` removal? 

if not needed anymore, please also remove the 
```
use Pimcore\Bundle\WebToPrintBundle\Model\Document\PrintAbstract;
use Pimcore\Bundle\WebToPrintBundle\Model\Document\Printpage;
```
as they are not used anymore in this file.

mattamon replied:
This is the only test connected to the web2print bundle and we decided to remove it so we do not have a dependency in the Core. 
We also decided that we do not move this one test to the bundle since the overhead for just one test would be too much. 
I will remove the uses.

kingjia90 reviewed (changes requested):
Thank you for fixing most of the stuff from the list in the last review.
I've found some other points, but overall looks and works fine to me.

kingjia90 reviewed (commented):
Please remove this checkbox and related translation/docs
https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/bundles/WebToPrintBundle/public/js/settings.js#L257-L266
which was used to define if the web2print should be activated or not 
https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/doc/Development_Documentation/03_Documents/02_Document_Types/15_Print_Documents.md?plain=1#L8

Also in the migration needs to check wheter the checkbox was previously marked by that config
https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/bundles/CoreBundle/src/Migrations/Version20221220152444.php#L37

kingjia90 reviewed (changes requested):
Some translations are still in the `admin_ext` and need to be extracted 
eg. 
https://github.com/pimcore/pimcore/blob/3eca81be035d22c7a3479d46003ad17ed8748778/bundles/AdminBundle/translations/admin_ext.en.yaml#L506-L533

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	bundles/AdminBundle/src/Controller/Admin/SettingsController.php

mattamon committed:
Move translations
Adding cleanup command
Adjust migration to remove or add config
Adjust installer to only remove permissions
Updated documentation

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221220152444.php:
```diff
@@ -0,0 +1,78 @@
+    public function up(Schema $schema): void
+    {
+        if(!SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle', 'pimcore')) {
+            SettingsStore::set('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle', true, 'bool', 'pimcore');
```
At least for `up`, something like
```suggestion
            SettingsStore::set('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle', $web_to_print_config['enableInDefaultView'], 'bool', 'pimcore');
```
if it was never activated before, on migration, it should stay deactivated.

For down, we don't know anymore, but, we could roughly check if there's any existing printpage? that could give an hint that it was previously checked, even though one could have not cleaned up any of those and still have web2print disabled.

mattamon replied:
This could potentially break the installer, since the permissions are already there.
If you upgrade to 11.x and then try to install the WebToPrintBundle (when it was not activated before) will break the installation since the permission is already in the table. 
We could check the `enableInDefaultView` and if it is false we also remove the permission?

In the down we then need to add the permission again if it is not there already.

kingjia90 replied:
ah, true, maybe we should change the installer to be more idempotent, instead of `insert`, to be an `upsert`.

The migration `up` here also need to consider if the `enableInDefaultView` is false and in that case, it should also remove the enum from `documents.types` and `documents_printpage` table

mattamon replied:
Hm, I am not so sure if we should do this. We cannot actual know if someone used this feature but disabled it for now. We would delete data with this migration, just because one setting was not set to true. This does not feel right to me.

mattamon replied:
We talked in the daily about it. We should not delete anything in the migration. Just do not mark it as installed and make the installer idempotent. We cannot trust this flag enough.

kingjia90 commented on bundles/WebToPrintBundle/src/Installer.php:
```diff
@@ -0,0 +1,122 @@
+    public function uninstall(): void
+    {
+        // Only remove permissions. Cleanup can be done by dev or command
+        $this->output->writeln("<info>Uninstalling only removes permissions. To clean up all documents and dependencies, please install the bundle again and run bin/console pimcore:documents:cleanup printpage printcontainer</info>");
```
```suggestion
        $this->output->writeln("<info>Uninstalling only removes permissions. To clean up all documents and dependencies, please run 'bin/console pimcore:documents:cleanup printpage printcontainer'</info>");
```
This doesn't require the web2print bundle to be active anymore, since it was suggested that the command would be useful to other types as well and that should reside in the core. 

Adding some quote to the actual command, otherwise could be hard to read where the command ends. I don't know if there's any tag like `<info>` that could highlight it. Some line breaks may also work.

mattamon replied:
The problem with this line is that it will never be shown somehow. I left it there because we also do it in the E-Commerce Framework. 

https://github.com/symfony/console/blob/6.2/Formatter/OutputFormatter.php#L73 The output formatter has some the info style but we could also change it to yellow e.g.

kingjia90 commented on lib/Twig/Extension/DocumentHelperExtensions.php:
```diff
@@ -53,18 +57,12 @@ public function getTests(): array
             new TwigTest('pimcore_document_snippet', static function ($object) {
                 return $object instanceof Document\Snippet;
             }),
+            new TwigTest('pimcore_document_print', static function ($object) {
+                return $object instanceof PrintAbstract;
+            }),
```
```suggestion
```

kingjia90 reviewed (changes requested):
found some other minor tweak to do while scanning through the last commit

mattamon committed:
Change uninstall info and remove TwigTest

mattamon committed:
Merge branch '11.x' into 12525-move-web2print-to-bundle

# Conflicts:
#	bundles/WebToPrintBundle/src/Processor/HeadlessChrome.php

mattamon committed:
Add use for Pimcore\Tool\Console

mattamon committed:
Make installer idempotent and enums unique

mattamon committed:
Make migration idempotent. Remove settings store setting in down
Only show enable bundle warning if bundle was marked as installed

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221220152444.php:
```diff
@@ -0,0 +1,86 @@
+        // restoring the enableInDefaultView might be with wrong value
+        SettingsStore::delete('BUNDLE_INSTALLED__Pimcore\\Bundle\\WebToPrintBundle\\PimcoreWebToPrintBundle','pimcore');
+        $settings = SettingsStore::get('web_to_print', 'pimcore_web_to_print');
+
```
In user permissions, web_to_print will have its on category (collapsible panel), should the migration down bring it back in the main list by clearing the `category` column (users_permission_definitions)?

mattamon committed:
Add Upgrading notes and warnings
Remove/add permissions in migration

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221220152444.php:
```diff
@@ -53,6 +48,13 @@ public function up(Schema $schema): void
+            if($enableBundle) {
+                $this->addSql("UPDATE `users_permission_definitions` SET `category` = 'Pimcore Web2Print Bundle' WHERE `key` = 'web2print_settings'");
+            } else {
+                $this->addSql('UPDATE `users` SET `permissions`=REGEXP_REPLACE(`permissions`, \'(?:^|,)web2print_settings(?:^|,)\', \'\') WHERE `permissions` REGEXP \'(?:^|,)web2print_settings(?:$|,)\'');
```
as mentioned in the call, this cleanup query should to be manually executed by devs, because we cannot trust the flag and it would be not possible to restore data with migration down.

kingjia90 commented on bundles/WebToPrintBundle/src/Installer.php:
```diff
@@ -0,0 +1,126 @@
+    public function uninstall(): void
+    {
+        // Only remove permissions. Cleanup can be done by dev or command
+        $this->output->writeln("<info>Uninstalling only removes permissions. To clean up all documents and dependencies, please run 'bin/console pimcore:documents:cleanup printpage printcontainer'</info>");
```
This doesn't show up but could be replaced with an `echo` for the moment. Maybe add some `#####` to highlight it a little bit

mattamon replied:
I actually just created a Consoleoutput like this.
` $output = new ConsoleOutput(OutputInterface::VERBOSITY_NORMAL, true);`

kingjia90 commented on doc/Development_Documentation/23_Installation_and_Upgrade/07_Updating_Pimcore/12_V10_to_V11.md:
```diff
@@ -1 +1,14 @@
+## CHANGES TO DO AFTER THE UPDATE! 
+
+### Web2Print
+If you do not use Web2Print functionality please make sure to run `bin/console pimcore:documents:cleanup printpage printcontainer` to clean up your system.
```
this command should be announced as a new feature in the upgrade notes and have a doc page about the usage.

Recommending that we should start filling this page with some more descriptive content and examples
https://github.com/pimcore/pimcore/edit/11.x/doc/Development_Documentation/22_Administration_of_Pimcore/05_Commandline_Interface.md
or alternatively somewhere under https://pimcore.com/docs/pimcore/11.0/Development_Documentation/Documents/index.html

mattamon committed:
Adding command documentation
Adding deleting queries instead of doing it in the migration
Adding consoleoutput on uninstall

mattamon committed:
Adding doc

kingjia90 commented on bundles/WebToPrintBundle/src/Installer.php:
```diff
@@ -0,0 +1,134 @@
+        $output = new ConsoleOutput(OutputInterface::VERBOSITY_NORMAL, true);
+        $output->writeln([
+            "\n\n<comment>Uninstalling only removes permissions. To clean up all documents and dependencies</comment>",
+            "<comment>Please run 'bin/console pimcore:documents:cleanup printpage printcontainer'</comment>",
```
```suggestion
            "<comment>Please run <options=bold>bin/console pimcore:documents:cleanup printpage printcontainer</></comment>",
```

In this way the command would appear in white, instead of all yellow, as follows:

![image](https://user-images.githubusercontent.com/6014195/219429893-a415005a-a724-49ee-aa07-004d830ea3ab.png)

kingjia90 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -13,6 +13,7 @@
 - [Documents] 
   - Removed `$types` property from `Pimcore\Model\Document`. Use `getTypes` method instead.
   - Removed `pimcore:document:types` from config. The types will be represented by the keys of the `type_definitions:map`
+  - Adding document cleanup command to clean up unused tables and document types, see [Documents](../../03_Documents/README.md#page_Cleanup-Documents-Types)
```
```suggestion
  - Added `pimcore:documents:cleanup` command to remove documents with specified types and drop the related document type tables, useful in the cases like the removal of headless documents or web2print page/containers after uninstallation,  see [Documents](../../03_Documents/README.md#page_Cleanup-Documents-Types)
```

kingjia90 reviewed (commented):
I will commit these final suggestions and looks good to be merged

kingjia90 committed:
Apply suggestions from code review
